File: src/main/java/no/nav/bidrag/beregn/felles/FellesBeregning.java
package no.nav.bidrag.beregn.felles;

import static java.util.Comparator.comparing;
import static java.util.stream.Collectors.toList;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import no.nav.bidrag.beregn.felles.bo.Periode;
import no.nav.bidrag.beregn.felles.bo.SjablonPeriode;
import no.nav.bidrag.beregn.felles.bo.SjablonPeriodeNavnVerdi;

public class FellesBeregning {

  // Mapper ut sjablonverdier til ResultatBeregning (dette for å sikre at kun sjabloner som faktisk er brukt legges ut i grunnlaget for beregning)
  protected List<SjablonPeriodeNavnVerdi> byggSjablonResultatListe(Map<String, BigDecimal> sjablonNavnVerdiMap,
      List<SjablonPeriode> sjablonPeriodeListe) {
    var sjablonPeriodeNavnVerdiListe = new ArrayList<SjablonPeriodeNavnVerdi>();
    sjablonNavnVerdiMap.forEach((sjablonNavn, sjablonVerdi) ->
        sjablonPeriodeNavnVerdiListe.add(new SjablonPeriodeNavnVerdi(hentPeriode(sjablonPeriodeListe, sjablonNavn), sjablonNavn, sjablonVerdi)));
    return sjablonPeriodeNavnVerdiListe.stream().sorted(comparing(SjablonPeriodeNavnVerdi::getNavn)).collect(toList());
  }

  private Periode hentPeriode(List<SjablonPeriode> sjablonPeriodeListe, String sjablonNavn) {
    return sjablonPeriodeListe.stream()
        .filter(sjablonPeriode -> sjablonPeriode.getSjablon().getNavn().equals(modifiserSjablonNavn(sjablonNavn)))
        .map(SjablonPeriode::getPeriode)
        .findFirst()
        .orElse(new Periode(LocalDate.MIN, LocalDate.MAX));
  }

  // Enkelte sjablonnavn må justeres for å finne riktig dato
  private String modifiserSjablonNavn(String sjablonNavn) {
    if (sjablonNavn.equals("UnderholdBeløp")) {
      return "Bidragsevne";
    }
    if (sjablonNavn.equals("BoutgiftBeløp")) {
      return "Bidragsevne";
    }
    if (sjablonNavn.startsWith("Trinnvis")) {
      return "TrinnvisSkattesats";
    }
    return sjablonNavn;
  }
}

File: src/main/kotlin/no/nav/bidrag/beregn/felles/FellesBeregning.kt
package no.nav.bidrag.beregn.felles

import no.nav.bidrag.beregn.felles.bo.Periode
import no.nav.bidrag.beregn.felles.bo.SjablonPeriode
import no.nav.bidrag.beregn.felles.bo.SjablonPeriodeNavnVerdi
import java.math.BigDecimal
import java.time.LocalDate

open class FellesBeregning {

    // Mapper ut sjablonverdier til ResultatBeregning (dette for å sikre at kun sjabloner som faktisk er brukt legges ut i grunnlaget for beregning)
    protected fun byggSjablonResultatListe(
        sjablonNavnVerdiMap: Map<String, BigDecimal>,
        sjablonPeriodeListe: List<SjablonPeriode>
    ): List<SjablonPeriodeNavnVerdi> {
        val sjablonPeriodeNavnVerdiListe = mutableListOf<SjablonPeriodeNavnVerdi>()
        sjablonNavnVerdiMap.forEach { (sjablonNavn: String, sjablonVerdi: BigDecimal) ->
            sjablonPeriodeNavnVerdiListe.add(
                SjablonPeriodeNavnVerdi(
                    periode = hentPeriode(sjablonPeriodeListe = sjablonPeriodeListe, sjablonNavn = sjablonNavn),
                    navn = sjablonNavn,
                    verdi = sjablonVerdi
                )
            )
        }

        return sjablonPeriodeNavnVerdiListe.sortedBy { it.navn }
    }

    private fun hentPeriode(sjablonPeriodeListe: List<SjablonPeriode>, sjablonNavn: String): Periode {
        return sjablonPeriodeListe
            .firstOrNull { it.sjablon.navn == modifiserSjablonNavn(sjablonNavn) }?.getPeriode() ?: Periode(LocalDate.MIN, LocalDate.MAX)
    }

    // Enkelte sjablonnavn må justeres for å finne riktig dato
    private fun modifiserSjablonNavn(sjablonNavn: String): String {
        return when (sjablonNavn) {
            "UnderholdBeløp" -> {
                "Bidragsevne"
            }

            "BoutgiftBeløp" -> {
                "Bidragsevne"
            }

            else -> if (sjablonNavn.startsWith("Trinnvis")) {
                "TrinnvisSkattesats"
            } else {
                sjablonNavn
            }
        }
    }
}
